### ISSUE TITLE
Trying to update from v3 to v4 but the removing the attributes in tfstate is hard to manage

### ISSUE THREAD

**Issue by gojanpaolo** [hashicorp/terraform-provider-google#15596]
Hello! We're currently trying to upgrade from v3 to v4 and noticed that there are some removed attributes like `bucket_policy_only` in `google_storage_bucket`. We deleted the attribute by manually editing the terraform state file. But whenever terraform refresh is ran (either by me or other developers), the attribute gets added back. Is there a general process on how to go about removing the attributes while still in v3 such that we'll have everything compatible to v4 when we actually update the versions used? Thanks!

**Comment by edwardmedia** [comment]
@gojanpaolo can you share a config you have and the steps for your upgrade?

**Comment by gojanpaolo** [comment]
@edwardmedia Thanks for the response! Which config do you mean? If you mean a .tf file then here's one definition we have
```
resource "google_storage_bucket" "tsg-1st-flink" {
  project = var.k8s_project
  name    = "tsg-1st-flink${var.environment_tag != "" ? "-${var.environment_tag}" : "" }"
  location = "US"
}
```

The steps we're doing is as follows:
1. Keep google version at 3.90.1
2. Run `terraform state pull > terraform.tfstate`
3. Manually remove `bucket_policy_only` attribute in the state file
4. Manually Increment state file's `serial`
5. Save the file
6. Run `terraform state push terraform.tfstate`
7. Run terraform refresh
8. Check the state.
  - The expectation is that `bucket_policy_only` should NOT be in there. 
  - The actual behavior is that `bucket_policy_only` was added back

**Comment by edwardmedia** [comment]
@gojanpaolo have you seen the expected behavior working on any other resources of any providers? I think this is a question more about the terraform, not specific the google provider. Look a closer look at the[ terraform refresh](https://developer.hashicorp.com/terraform/cli/commands/refresh), I noticed below comment. Does it make sense?

```
This won't modify your real remote objects, but it will modify the [Terraform state](https://developer.hashicorp.com/terraform/language/state).
```

**Comment by gojanpaolo** [comment]
I see. Is there a general process on how to go about updating from google v3 to v4 w/ respect to the removed attributes/fields? Or is the process something like:
1. Update google provider to v4
i. This means `terraform refresh` (or `terraform plan / apply` without `-refresh=false`) will be broken until Step 2 is completed
2. Update terraform state manually to delete all removed attributes

**Comment by edwardmedia** [comment]
@gojanpaolo is [this](https://developer.hashicorp.com/terraform/cli/commands/init#upgrade) something you were looking for?

```
terraform init --upgrade
```

**Comment by gojanpaolo** [comment]
No. That upgrades the version. But then if we're coming from v3, then the terraform state is incompatible with the new version v4 because of the removed attributes (e.g. `bucket_policy_only` in `google_storage_bucket`).

**Comment by edwardmedia** [comment]
@gojanpaolo sorry I don't get what exactly you want to achieve. Why you want to manually modify the state file? Have you seen a use case where other people did things like that?

**Comment by gojanpaolo** [comment]
> Why you want to manually modify the state file?
Because there are attributes added from v3 that are now removed in v4. Those attributes does not seem to be automatically handled by terraform / google provider. And we get the following error when we try to do a `terraform refresh` when we update to v4

```
Error decoding "module.dev.module.k8s-tsg-1st.module.flink[0].google_storage_bucket.tsg-1st-flink" from previous state: unsupported attribute "bucket_policy_only"
```

**Comment by edwardmedia** [comment]
@gojanpaolo I guess after you upgraded the provider version from v3 to v4 with `terraform init --upgrade`, things do not work as you wanted so you need to manually fix them? 

Do you have a specific config that you can show us? Please detail the steps how to repro and where to see the issue. 

```
Those attributes does not seem to be automatically handled by terraform / google provider.
```

**Comment by gojanpaolo** [comment]
I think something like this would work but I'll see when I can find the time to verify

1. Use terraform google provider v3
2. Apply the following resource
```
resource "google_storage_bucket" "tsg-1st-flink" {
  project = var.k8s_project
  name    = "test-bucket-google-provider-issue-15596"
  location = "US"
}
```
3. Run `terraform pull` and check that the state has `bucket_policy_only`
4. Update to terraform google provider v4
5. Run `terraform refresh`

Expected behavior: `terraform refresh` on v4 should run successfully
Actual behavior: (TODO: verify this) Errors out with `Error decoding "..." from previous state: unsupported attribute "bucket_policy_only"`

**Comment by edwardmedia** [comment]
@gojanpaolo for step 3, I guess it is `terraform state pull`? Why do you want to do that? And what is the purpose of doing step 5 `terraform refresh`? Where does your state file reside?

**Comment by gojanpaolo** [comment]
Ah yes `terraform state pull`. It's only to check that the deprecated attribute (`bucket_policy_only`) is there. It's not necessary to actually reproduce the issue.

Step 5 demonstrates that it's broken because of the removed attribute in v4.

Our actual state file resides in a storage bucket.

**Comment by rileykarson** [comment]
I've tested an upgrade locally and confirmed it works for me: https://gist.github.com/rileykarson/d7951e96996dfa3305f4e5023fa6d7d6. This is probably an issue w/ remote state backends.

**Comment by ggtisc** [comment]
Hi @gojanpaolo 

If you have differences between you have differences between your terraform project and your GCP project you just need to run a `terraform refresh` command. Otherwise if you are still having issues please open a new ticket sharing your code and the steps to replicate your issue. For sensitive data you could use examples like:

project = "my-project"
user = "my-user@my-domain.com"
org_id = 1234567890
